Better docs for completions

serras opened the pull request:
This PR makes completion search for documentation using the [`getDocs` function](https://downloads.haskell.org/~ghc/latest/docs/html/libraries/ghc-8.8.1/GHC.html#v:getDocs).

serras commented:
There's something very strange: the new search for documentation works when I run `ghcide` in VS Code, but tests fail, and I cannot find a reason why.

serras commented:
Once this is solved, the next step is [reading the Haddock markup](http://hackage.haskell.org/package/pandoc-2.9/docs/Text-Pandoc-Readers-Haddock.html) and [producing Markdown](http://hackage.haskell.org/package/pandoc-2.9/docs/Text-Pandoc-Writers-Markdown.html) with the help of `pandoc`.

serras commented:
I'll do it on Monday, once I'm back in action.

For the Haddock to Markdown conversion: do you know of any alternative to
Markdown? Would it be OK to depend on Haddock-the-library for parsing the
docs and then turning them to Markdown by hand?

El vie., 3 ene. 2020 9:59, Moritz Kiefer <notifications@github.com>
escribió:

> *@cocreature* commented on this pull request.
>
> Sorry for the slow response.
>
> It would be great if we could separate the removal of the FromJSON
> instances from the documentation fixes and make two PRs to make this easier
> to review.
>
> As for the changes, I am a bit hesitant to add a dependency on pandoc.
> It’s a pretty large dependency for the relatively small task of converting
> haddock to markdown.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/digital-asset/ghcide/pull/288?email_source=notifications&email_token=AACLQVUGIUTFCQQEQWRM4WTQ3345JA5CNFSM4J57JOHKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQSVHBI#pullrequestreview-337990533>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACLQVSQPRY7AHORJDFAPI3Q3345JANCNFSM4J57JOHA>
> .
>

cocreature commented:
Depending on the haddock library would be fine. Alternatively, we could also go for something a bit more hacky and just do a bunch of search and replace statements to approximate the conversion.

serras commented:
The part about removing unused JSON instances is now #305. In this PR I have reverted those changes, and also moved from Pandoc to a hand-written Haddock-to-Markdown conversion.

It should now be ready for review. Still, I would suggest to merge #305 first, and then work on merging both branches as part of this PR, since there are some conflicts in the tests between them.

cocreature commented on test/exe/Main.hs:
```diff
@@ -1061,31 +1061,33 @@ completionTests
       , _kind = kind
-      , _detail = Just "Prelude"
-      , _documentation = Just (CompletionDocMarkup (MarkupContent {_kind = MkMarkdown, _value = ""}))
+      , _detail = (":: " <>) <$> ty
```
This should also check `optNewColonConvention`.

serras replied:
I am not sure how to do this, since I don't know where the tests set this flag.

cocreature replied:
Oh I missed that this is a test. Fine to leave it in that case since the test will always use double colons.

cocreature commented on test/exe/Main.hs:
```diff
@@ -1061,31 +1061,33 @@ completionTests
   where
-    complItem label xdata kind = CompletionItem
+    dropDocs :: CompletionItem -> CompletionItem
+    dropDocs ci = ci { _documentation = Nothing }
```
Instead of dropping docs, why don’t we check them in the tests? That seems much more useful.

serras replied:
Since the tests would get very big, I'm checking that the first 100 characters match.

serras commented on src/Development/IDE/Core/Completions.hs:
```diff
@@ -187,8 +195,8 @@ occNameToComKind oc
 mkCompl CI{origName,importedFrom,thingType,label,isInfix,docs} =
-  CompletionItem label kind (Just $ maybe "" (<>"\n") typeText <> importedFrom)
-    (Just $ CompletionDocMarkup $ MarkupContent MkMarkdown $ T.intercalate sectionSeparator docs)
+  CompletionItem label kind ((":: "<>) <$> typeText)
```
Done in the last commit

cocreature commented:
Looks like GHC 8.4 is failing due to `-Werror` and GHC 8.8 is failing due to `haddock-library` not being in stackage which can hopefully be solved by adding that to `stack88.yaml`.

serras commented:
I'll try to fix both problems tonight. Thanks for the review!

El lun., 6 ene. 2020 14:43, Moritz Kiefer <notifications@github.com>
escribió:

> Looks like GHC 8.4 is failing due to -Werror and GHC 8.8 is failing due
> to haddock-library not being in stackage which can hopefully be solved by
> adding that to stack88.yaml.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/digital-asset/ghcide/pull/288?email_source=notifications&email_token=AACLQVSR6DDRSA5MMJ76643Q4MYRJA5CNFSM4J57JOHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIFPGEY#issuecomment-571142931>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACLQVXODH2NNKNMBDYSH5LQ4MYRJANCNFSM4J57JOHA>
> .
>

serras commented:
Finally everything is green!

cocreature commented on test/exe/Main.hs:
```diff
@@ -1243,7 +1243,7 @@ completionTests
         compls <- getCompletions docId (Position 1 7)
         liftIO $ map dropDocs compls @?= 
           [complItem "head" (Just CiFunction) (Just "[a] -> a")
-                     "*Defined in 'Prelude'*\n*\t*\t*\n\n\nExtract the first element of a list, which must be non-empty."]
```
This doesn’t actually check anything about the docs. The “Defined in Prelude” part is completely independent of this PR (although this PR slightly changes the wording). Let’s please check that the docs actually work. Some CPP to not check this on older GHCs is totally fine.

serras replied:
New tests are in!
